File: app/src/main/java/com/dpis/module/templates/presentation/TemplateWorkspacePresentationController.java
package com.dpis.module.templates.presentation;

import android.content.Context;

import com.dpis.module.templates.TemplateEditorDraft;

import java.util.LinkedHashSet;
import java.util.Set;
import com.dpis.module.ui.ConfigEditorDestination;

/** Owns template-list refresh timing so Compose never reads template stores during recomposition. */
public final class TemplateWorkspacePresentationController {
    public interface Listener {
        void onStateChanged(TemplateWorkspacePresentation.State state);
    }

    private final Context context;
    private final TemplateWorkspacePresentation.Actions actions;
    private final Set<Listener> listeners = new LinkedHashSet<>();
    private TemplateWorkspacePresentation.State state;

    public TemplateWorkspacePresentationController(
            Context context,
            TemplateWorkspacePresentation.Actions actions,
            String initialQuery
    ) {
        this.context = context;
        this.actions = actions;
        this.state = TemplateWorkspacePresentation.create(context, initialQuery, actions);
    }

    public TemplateWorkspacePresentation.State state() {
        return state;
    }

    /** Republishes one complete editor session snapshot; callers may not drop route or draft state. */
    public void refresh(
            String query,
            TemplateWorkspacePresentation.DetailKind nextDetailKind,
            String nextDetailTemplateId,
            com.dpis.module.ui.ConfigEditorDestination editorDestination,
            TemplateEditorDraft globalPrefillDraft,
            TemplateEditorDraft quickTemplateDraft,
            TemplateWorkspacePresentation.ApplyConfirmation applyConfirmation
    ) {
        TemplateWorkspacePresentation.DetailKind detailKind = nextDetailKind != null
                ? nextDetailKind
                : TemplateWorkspacePresentation.DetailKind.NONE;
        state = TemplateWorkspacePresentation.create(
                context,
                query,
                actions,
                detailKind,
                nextDetailTemplateId,
                editorDestination,
                globalPrefillDraft,
                quickTemplateDraft,
                applyConfirmation
        );
        for (Listener listener : new LinkedHashSet<>(listeners)) {
            listener.onStateChanged(state);
        }
    }

    public void addListener(Listener listener) {
        if (listener == null) return;
        listeners.add(listener);
        listener.onStateChanged(state);
    }

    public void removeListener(Listener listener) {
        listeners.remove(listener);
    }
}

File: app/src/main/java/com/dpis/module/templates/presentation/TemplateWorkspacePresentationController.kt
package com.dpis.module.templates.presentation

import android.content.Context
import com.dpis.module.templates.TemplateEditorDraft
import com.dpis.module.ui.ConfigEditorDestination
import java.util.LinkedHashSet

/** Owns template-list refresh timing so Compose never reads template stores during recomposition. */
class TemplateWorkspacePresentationController(
    private val context: Context,
    private val actions: TemplateWorkspacePresentation.Actions,
    initialQuery: String,
) {
    fun interface Listener {
        fun onStateChanged(state: TemplateWorkspacePresentation.State)
    }

    private val listeners = LinkedHashSet<Listener>()
    private var snapshot: TemplateWorkspacePresentation.State =
        TemplateWorkspacePresentation.create(context, initialQuery, actions)

    fun state(): TemplateWorkspacePresentation.State = snapshot

    /** Republishes one complete editor session snapshot; callers may not drop route or draft state. */
    fun refresh(
        query: String?,
        nextDetailKind: TemplateWorkspacePresentation.DetailKind?,
        nextDetailTemplateId: String?,
        editorDestination: ConfigEditorDestination?,
        globalPrefillDraft: TemplateEditorDraft?,
        quickTemplateDraft: TemplateEditorDraft?,
        applyConfirmation: TemplateWorkspacePresentation.ApplyConfirmation?,
    ) {
        val detailKind = nextDetailKind ?: TemplateWorkspacePresentation.DetailKind.NONE
        snapshot = TemplateWorkspacePresentation.create(
            context,
            query,
            actions,
            detailKind,
            nextDetailTemplateId,
            editorDestination ?: ConfigEditorDestination.MAIN,
            globalPrefillDraft,
            quickTemplateDraft,
            applyConfirmation,
        )
        for (listener in LinkedHashSet(listeners)) {
            listener.onStateChanged(snapshot)
        }
    }

    fun addListener(listener: Listener?) {
        if (listener == null) return
        listeners.add(listener)
        listener.onStateChanged(snapshot)
    }

    fun removeListener(listener: Listener?) {
        listeners.remove(listener)
    }
}
